test/config-luatest/sources_test.lua

```diff
@@ -69,6 +71,104 @@ g.test_source_env = function()
+            env = {
+                TT_LOG_LEVEL = 'info',
+                TT_MEMTX_MEMORY = 1000000,
+            },
+        },
+        {
+            name = 'env default',
+            env_var_suffix = 'default',
+            env = {
+                TT_LOG_LEVEL_DEFAULT = 'info',
+                TT_MEMTX_MEMORY_DEFAULT = 1000000,
+            },
+        },
+    }
+    local opts = {nojson = true, stderr = false}
+    for _, case in ipairs(cases) do
+        local comment = ('case: %s'):format(case.name)
+        local args = {'main.lua', case.env_var_suffix}
+        local res = justrun.tarantool(dir, case.env, args, opts)
+        t.assert_equals(res.exit_code, 0, comment)
+        t.assert_equals(json.decode(res.stdout), exp, comment)
+    end
+end
+
+-- Verify priority of configuration sources.
+--
+-- 1. env (TT_*)
+-- 2. file
+-- 3. env default (TT_*_DEFAULT)
+--
+-- Several string options from the instance config are chosen for
+-- testing purposes, their meaning is irrelevant for the test.
+--
+-- The table below shows where the given option is set (which
+-- source defines it) and what we expect as a result.
+--
+-- | option              | env | file | env default | result      |
+-- | ------------------- | --- | ---- | ----------- | ----------- |
+-- | process.title       |     |  +   |      +      | file        |
+-- | log.file            |  +  |  +   |             | env         |
+-- | log.pipe            |  +  |      |      +      | env         |
+-- | log.syslog.identity |     |      |      +      | env default |
+g.test_sources_priority = function(g)
+    local dir = treegen.prepare_directory(g, {}, {})
+    local config = {
+        credentials = {
+            users = {
+                guest = {
+                    roles = {'super'},
+                },
+            },
+        },
+        iproto = {
+            listen = 'unix/:./{{ instance_name }}.iproto',
+        },
+        process = {
+            title = 'from file',
+        },
+        log = {
+            file = 'from file',
+        },
+        groups = {
+            ['group-001'] = {
+                replicasets = {
+                    ['replicaset-001'] = {
+                        instances = {
+                            ['instance-001'] = {},
+                        },
+                    },
+                },
+            },
+        },
+    }
+    local config_file = treegen.write_script(dir, 'config.yaml',
+                                             yaml.encode(config))
+    local opts = {
+        config_file = config_file,
+        chdir = dir,
+        env = {
+            TT_PROCESS_TITLE_DEFAULT = 'from env default',
+            TT_LOG_FILE = 'from env',
+            TT_LOG_PIPE = 'from env',
+            TT_LOG_PIPE_DEFAULT = 'from env default',
+            TT_LOG_SYSLOG_IDENTITY_DEFAULT = 'from env default',
+        },
+    }
+    g.server = server:new(fun.chain(opts, {alias = 'instance-001'}):tomap())
+    g.server:start()
```

Comment:
Needs `g.server:stop()` in an after_test/after_each hook to stop the tarantool process if the test case fails.

Reply:
Added the `after_each()` hook.